### DISCUSSION TITLE
Formula/Cask to Automate Machine Setup

### DISCUSSION THREAD

**Discussion by danielburn5** [Homebrew/discussions#3057] [Writing Formulae/Casks] [answered]
Whenever my company has new engineers join they need to setup their laptop which includes a bunch of `brew` installs. At the moment we have a `BrewFile` which they can use to install things. I was wondering if this might be better suited as a formula or cask, so that it can benefit from being installed / removed / updated in one hit.

However, I'm not sure if this really fits the mould for formulas or casks, given that the content of said formula/cask would be purely dependencies (ie the formula would just install other formulas and/or casks, but wouldn't be installing anything new. I guess it would also mean that we'd effectively be dictating which versions to install to everyone as part of formula/cask, whereas with the existing BrewFile approach, each installed package can be managed independently.

Figured I'd open a discussion and get some input from other who know more about Homebrew than I 🙂

**Comment by SMillerDev** [comment]
A formula with dependencies and a brew file could accomplish most of the same things but I use Brewfiles for my work for 2 reasons.
1. The DSL is easier, allowing more colleagues to contribute.
1. You can remove any additional software easily (I believe it's `brew bundle cleanup`) which isn't happening if you uninstall a formula/cask.

**Comment by carlocab** [comment, marked as answer]
A `Brewfile` is pretty much what you seem to need here.

If you want to install everything in your `Brewfile`: `brew bundle install`

If you want to remove everything not in your `Brewfile`: `brew bundle cleanup --force` (this won't remove dependencies of formulae/casks in your `Brewfile`)

If you want to upgrade everything you have installed: `brew upgrade`

You can use `brew bundle cleanup` to also remove everything by removing your `Brewfile` then doing `brew bundle cleanup --force`.

Finally, if you have a Homebrew installation you want to port to a different machine, you can use `brew bundle dump` to generate a `Brewfile` for all the formulae you've installed on-request.

**Comment by danielburn5** [comment]
Thanks guys. `BrewFile` it is!
